buildtest/modules/util.py

```diff
@@ -1,48 +1,156 @@
-import subprocess
+from buildtest.tools.system import BuildTestCommand
 
 
 def get_all_collections():
-    """Return all module collection """
+    """Get all Lmod user collections that is retrieved by running ``module -t savelist``.
+
+     :return: Return all module collections
+     :rtype: list
+     """
+
     collections = "module -t savelist"
-    ret = subprocess.Popen(
-        collections, shell=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE
-    )
-    out = ret.communicate()[1]
-    tmp = out.decode("utf-8").split()
+    cmd = BuildTestCommand()
+    cmd.execute(collections)
+    # The output is captured in STDERR stream.
+    out = cmd.get_error().split()
 
-    return tmp
+    return out
 
 
 class ModuleCollection:
+    """Class declaration of ModuleCollection."""
+
     def __init__(self, collection):
+        """Initializer method of ModuleCollection class.
+
+        :param collection: name of module collection
+        :type collection: str
+        """
+
+        # raise TypeError exception if collection is not a string type since that is required
+        # when working with module collection
+        if not isinstance(collection, str):
+            raise TypeError(f"Type Error: {collection} is not of type string")
+
         self.collection = collection
         self.module_cmd = f"module restore {self.collection}"
 
     def test_collection(self):
-        ret = subprocess.Popen(
-            self.module_cmd, shell=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE
-        )
-        ret.communicate()
-        return ret.returncode
+        """Test the module collection by running ``module restore <collection>``.
+
+        :return: return code of ``module restore`` command
+        :rtype: int
+        """
+
+        cmd = BuildTestCommand()
+        cmd.execute(self.module_cmd)
+        return cmd.returnCode()
 
     def get_command(self):
+        """ Get the module command used to restore a collection
+
+        :return: Return the actual command to restore a collection.
+        :rtype: str
```

Comment:
What is an example of this command? I'm not super familiar with Lmod / collections.

Reply:
See https://github.com/shahzebsiddiqui/buildtest-framework/blob/module_api_v2.0/examples/collection.py#L14 this will return the module collection command. This is useful during test creation. Look at https://buildtest.readthedocs.io/en/latest/build_subcommand/module_collections.html#building-test-with-lmod-user-collection-buildtest-build-collection this example shows a CUDA collection loaded in test script to test a saxpy code. This is a way for users to save module collection and specify them in YAML file without having to worry about actual module names. 

Once you read https://lmod.readthedocs.io/en/latest/010_user.html#user-collections you will get idea why user collection are important.